osh/prompt.py

```diff
@@ -117,9 +117,11 @@ def _ReplaceBackslashCodes(self, tokens):
 
       elif id_ == Id.PS_LBrace:
         non_printing += 1
+        ret.append('\x01')
 
       elif id_ == Id.PS_RBrace:
         non_printing -= 1
```

Comment:
Yeah I think it makes sense to check if non_printing is not 0 at the end of the function and render an error?

What does bash do?  (If I had to guess, it causes problems without producing a useful error...)

Reply:
This is interesting: if PS1 is `> \[\[ >`, then when you go off the line, *it wraps properly*, but then a character later goes to the start of the second line and starts overwriting the line. I would have expected one or the other, seeing both is strange.

Reply:
So you think we should catch the error here instead of in `comp_ui.py`? That sounds like a plan, I can move my invalid inputs tests over to that test file

Reply:
Yeah I think validating in prompt.py makes sense.  Then just return an error string.  And then the length calculation will work fine on the error string -- it won't have unbalanced `\x01` and `\x02`, because you will have caught the unbalanced `\[` and `\]`.

Reply:
I added this, it will show `<error parsing prompt>` but still let you use the shell without throwing an exception.

![image](https://user-images.githubusercontent.com/23638587/54683687-a6a1cb00-4ae8-11e9-89eb-9a475df6c652.png)